Add Internet Archive daily item tasks

varun-magesh opened the pull request:
This PR adds three methods created to handle uploads to Internet Archive 'daily items' --- buckets that store all of the perma links made in a day. It also adds some settings needed for these methods.

## Description
The three methods are:
1. `upload_to_internet_archive_daily_item(link_guid)`, similar to `upload_to_internet_archive` that puts a link and its metadata into a daily item instead.
2. `delete_from_internet_archive_daily_item(link_guid)`, a replacement for `delete_from_internet_archive` that looks for a link in its daily item and deletes it from there instead.
3. `derive_internet_archive_daily_item(iso_date_str)`, which manually runs the derive process for an IA item --- this process is normally skipped for individual uploads.

## Deployment
`upload_to_internet_archive_daily_item` can be run at any time in the place of `upload_to_internet_archive`. Wherever the latter is currently run, the former can go instead. `derive_internet_archive_daily_item` should be run once at the end of each day for the previous day --- e.g. at `2022-07-26T23:59` we run derive for `2022-07-25`, because all of that day's files should be uploaded by now with the 24 hour delay.

varun-magesh committed:
Create bulk_upload task

varun-magesh committed:
Bulk_upload_to_internet_archive method complete

varun-magesh committed:
Add upload_to_internet_archive_daily_item

varun-magesh committed:
Create ia_daily_identifier names

varun-magesh committed:
Added ia daily item delete task

varun-magesh committed:
Added derive ia item task

varun-magesh committed:
Remove bulk_upload_to_internet_archive

varun-magesh committed:
Remove bulk_upload, add some documentation

bensteinberg commented on perma_web/requirements.in:
```diff
@@ -95,4 +95,6 @@ watchdog                                # watch for file changes
\ No newline at end of file
+sentry-sdk
+
+dict2xml
```
This is tiny, but I think I would put this somewhere up higher in `requirements.in`, either in the general section, or maybe better in link preservation under `internetarchive`?

varun-magesh committed:
Add upload_all_to_internet_archive_daily_item to fabfile

varun-magesh committed:
Replace post_processing tasks with daily_item

varun-magesh committed:
Raise error if login session is unsecure

varun-magesh committed:
Change order of dict2xml import

varun-magesh commented:
All changes we discussed in dev meeting 2022-07-27 have been made!

varun-magesh committed:
Add derive task to crontab

bensteinberg commented:
I may be missing this, but I don't think the new `upload_all_to_internet_archive_daily_item()` observes `INTERNET_ARCHIVE_MAX_UPLOAD_SIZE` like `upload_all_to_internet_archive()` does in

https://github.com/harvard-lil/perma/blob/6f9dcf4755cccfe6722737d7bfafcb325cd4bd88/perma_web/perma/tasks.py#L1567

(For others seeing this, WARCs larger than or equal to `INTERNET_ARCHIVE_MAX_UPLOAD_SIZE` are uploaded separately via a cron job.)

varun-magesh committed:
Fixed lints ant typos

varun-magesh commented:
> I may be missing this, but I don't think the new `upload_all_to_internet_archive_daily_item()` observes `INTERNET_ARCHIVE_MAX_UPLOAD_SIZE` like `upload_all_to_internet_archive()` does in
> 
> https://github.com/harvard-lil/perma/blob/6f9dcf4755cccfe6722737d7bfafcb325cd4bd88/perma_web/perma/tasks.py#L1567
> 
> (For others seeing this, WARCs larger than or equal to `INTERNET_ARCHIVE_MAX_UPLOAD_SIZE` are uploaded separately via a cron job.)

I think/hope it still does --- https://github.com/varun-iyer/perma/blob/93671726dd3ded92e90cd72343c98fa79c12e33c/perma_web/perma/tasks.py#L1618 the logic is still in there, but in a slightly different place

rebeccacremona commented on perma_web/fabfile/dev.py:
```diff
@@ -315,24 +315,33 @@ def test_internet_archive():
-    for link in links:
+
+@task
+def upload_all_to_internet_archive():
```
Hmm, I don't even know why we have a fab task for uploading things to internet archive... I know I have never used it; `git blame` suggests this function hasn't been altered for 5 or 6 years.

I think we can safely remove `fab.dev upload_all_to_internet_archive`, as well as the new `upload_all_to_internet_archive_daily_item` and helper function: the Celery tasks have us covered!

rebeccacremona commented on perma_web/perma/tasks.py:
```diff
@@ -1478,6 +1480,41 @@ def cache_playback_status(link_guid):
+    ]
+    if not item.exists:
+        logger.info(f"Daily item {link.ia_daily_item_identifier} not present in IA: skipping.")
+        return False
```
I think before `return`ing here, we probably need to update the status of the link, so that we don't try deleting it again in the next hour's batch. Ha, looks like that's a bug in the existing task, just replicated here.

varun-magesh replied:
Hmm, would we want the status set to failed or deleted? I can see a case for both here.

rebeccacremona commented on perma_web/perma/tasks.py:
```diff
@@ -1478,6 +1480,41 @@ def cache_playback_status(link_guid):
+            )
+        except Exception:
+            link.internet_archive_upload_status = 'deletion_incomplete'
+            logger.exception(f"Link {link.guid}: attempt to delete file {f} from Internet Archive failed:")
```
After the files are deleted, do you know if the `derive` process runs again automatically, removing all references to the deleted content from the derived files, or if we'll have to manually trigger `derive`?

varun-magesh replied:
On investigation, I think the the `cascade_delete` flag needs to be set here (https://archive.org/services/docs/api/internetarchive/api.html?highlight=delete#deleting). I'm not sure, however, what the difference is between cascading deletion and actually re-triggering derive --- they might be the same, or totally different. I know that re-triggering derive would redo the derive for all files in the item, not just this one, so we should test this to see how the behavior of cascade_delete and derive differ. If cascade_delete has identical behavior then we should avoid re-queueing derive.

varun-magesh replied:
Oh the cascade delete flag is set --- this is a part of our larger picture then.

rebeccacremona commented on perma_web/perma/tasks.py:
```diff
@@ -1658,6 +1712,112 @@ def upload_to_internet_archive(link_guid):
+def upload_to_internet_archive_daily_item(link_guid):
+    """
+    This method is functionally similar to `upload_to_internet_archive`.
+    It is a drop-in replacement: whereever upload_to_internet_archive is run
```
`whereever`->`wherever`

rebeccacremona commented on perma_web/perma/tasks.py:
```diff
@@ -1658,6 +1712,112 @@ def upload_to_internet_archive(link_guid):
+            retries_sleep=5,
+            queue_derive=False
+        )
+        response_list[0].raise_for_status()
```
Is this another good spot for the `[response]` syntax? Is the response_list ever in fact a list? (Maybe if there are retries? I'm not sure how a person could check that...) If it is in fact a list, we might not want `response_list[0]` to raise, we might want the last element in the list instead.

I see that that's just copying the syntax and behavior of the existing function; probably, like the thing I pointed out in `delete...`, an existing bug.

varun-magesh replied:
It returns a list of request response objects: https://archive.org/services/docs/api/internetarchive/api.html#internetarchive.upload, but cryptic on why there's a list/what each element of the series represents. Let me take a look at the source and see what the right behaviour should be.

varun-magesh replied:
Looks like there's a response for each file uploaded: https://github.com/jjjake/internetarchive/blob/83c9cd88a2d87cd52972976c5d49994297f99411/internetarchive/item.py#L1129, of which we only have one. Still, I think more complete behavior is to check that all reponses are not failure states.

rebeccacremona reviewed (changes requested):
This looks great!! Thank you so much for working on this project.

I left a couple of questions/comments, which are less about your code, and more about Qs raised from seeing this process with fresh eyes.

Two additional thoughts.

### derive

I think we are going to need to be a little more nuanced about when we call the `derive` process: 

- if a person toggles a Perma Link to private, we may need to run `derive` again after we delete it (as mentioned in the code; not sure)
- if a person toggles a private link to public on an arbitrary day after its creation (a common occurrence), it will get uploaded to its correct daily item via this code, but we won't know to re-run `derive` on that daily item.
- if all the items for the previous day have not yet been uploaded to the daily item when the daily `derive` cron task runs... maybe because there are still large warcs to be uploaded via the independent process, maybe because IA had an outage, maybe because we are behind processing the queue (which isn't infrequent), then we will `derive` too soon. 
- if the `derive` call fails, we have no mechanism for recording that failure and triggering it again later

I think this probably means that we want to start tracking which daily items need to be `derive`ed or re-`derived` in the database. Maybe we should add a `InternetArchiveItem` model... 

### For convenience

I think it would be convenient to be able to merge and deploy this code without necessarily turning it on at the same time. Could you set it up so that the new tasks are behind a feature flag, `settings.INTERNET_ARCHIVE_DAILY_ITEM = False` by default?

I know your time at LIL is coming to a close 😢; if there's not time to address any of this, I will pick up it up!! With the feature flag, we can merge without it being complete.

varun-magesh committed:
Remove IA Fab tasks

varun-magesh committed:
Fixed deletion status update bug

varun-magesh committed:
Check responses more completely

varun-magesh committed:
Add upload derive status

varun-magesh committed:
Added daily_item feature flag and internet_archive_identifier

The feature flag controls whether new items are uploaded to daily items
or link items by default (the other option is still easily available).
The identifier tracks where items have been uploaded --- the daily item,
or a link item.

What still needs to be done is handling/populating nulls on the daily
item, and adding derive_all logic.

rebeccacremona committed:
Remove old IA code from fabfile.

rebeccacremona committed:
Use a string for simplicity; account for settings being lazy.

rebeccacremona committed:
Use @property syntax for convenience.

rebeccacremona committed:
Whitespace.

rebeccacremona committed:
Use the extracted helper function.

rebeccacremona committed:
Typo.

rebeccacremona committed:
Add IA identifier field to the Django admin.

rebeccacremona committed:
Redo migration w/blank=True so admin form saves w/o an identifier

rebeccacremona commented:
This is getting close!!

Obstacles to merging:

- If merged, `link.internet_archive_identifier` must be populated for us to subsequently delete the item from IA, but, there is no data migration to populate the field for already-uploaded items, or an equivalent work around

- Re-uploading an (old-style) link item after deleting it is not working; I need to test at a slower pace, but, it may not be working with the production code either: 400 during the metadata refresh (fixed by not attempting to alter "collection")... and 200 that nevertheless seems to be a no-op when actually re-uploading the warc, or attempting to upload a testfile foo.txt to the existing item... I'll keep looking; there is a possibility that my items (or even possibly my personal account) are in a bad state after attempting to upload/delete/upload too quickly.

- `link.internet_archive_identifier = link.ia_link_item_identifier` is set at the very end of a successful upload. in our experience, partial uploads frequently succeed. We should think carefully about when we want to set that field. And think a little more about whether this field meets our needs long term, or only serves to indicate which strategy (`daily` or `link`) was used during any upload attempts... in which case, we might be better served by a Boolean field w/a default value or similar.

- All upload and delete functions should be robust against `link.internet_archive_identifier` = None; they should not rely on a function higher up in the call chain to prevent them from being called on a link with `link.internet_archive_identifier` = None. (This is especially important because the IA library will happily interact with an object `internetarchive.get_item(None)`.

I will keep chipping away at it!

rebeccacremona committed:
Whitespace removed to resolve merge conflct.

rebeccacremona committed:
Add decorator that makes the functions celery tasks.

rebeccacremona commented:
> Re-uploading an (old-style) link item after deleting it is not working

Update: okay, with a small tweak, re-uploading does indeed seem to work as desired. A person has to be careful when testing: it's not good to mess around with an Item when it has tasks pending or in progress (visible via the GUI if logged in as the creator at https://archive.org/history/{identifier} or via internetarchive.get_item(identifier).tasks... which must be fetched fresh for updates). When you upload a file after the fact, it doesn't (evidently) appear in the details GUI or in item.files until the whole queue of tasks is processed... which, if it includes a `derive` task, may take quite some time indeed. If a task errors, which can also happen if uploading and deleting and reuploading quickly, I'm not sure what happens: I think it might actually hang pending admin intervention, but I can't confirm.

Testing with `queue_derive=False` in the test collection, and waiting for tasks to finish before acting, upload-delete-reupload worked fine (with the small tweak).

And... test with `queue_derive=True` subsequently worked. Cool. So this is much closer to fine than I thought 🙌

rebeccacremona committed:
Don't attempt to set collection, when re-uploading a deleted item.

rebeccacremona commented:
This PR set the foundation for our current strategy. Thank you so much for all the work!
